File: hard/robot_movements.r
robot <- function(f, x, y) {
  if (x == 3 && y == 3) {
    return(1)
  }
  ret <- 0
  if (x > 0 && !((x-1 + 4*y) %in% f)) {
    ret = ret + robot(append(f, x-1 + 4*y), x-1, y)
  }
  if (y > 0 && !((x + 4*(y-1)) %in% f)) {
    ret = ret + robot(append(f, x + 4*(y-1)), x, y-1)
  }
  if (x < 3 && !((x+1 + 4*y) %in% f)) {
    ret = ret + robot(append(f, x+1 + 4*y), x+1, y)
  }
  if (y < 3 && !((x + 4*(y+1)) %in% f)) {
    ret = ret + robot(append(f, x + 4*(y+1)), x, y+1)
  }
  ret
}

cat(robot(c(0), 0, 0), sep="\n")

File: hard/robot_movements.rb
def robot(f, x, y)
  return 1 if x == 3 && y == 3
  ret = 0
  if x > 0 && f & (1 << (x - 1 + 4 * y)) == 0
    ret += robot(f | (1 << (x - 1 + 4 * y)), x - 1, y)
  end
  if y > 0 && f & (1 << (x + 4 * (y - 1))) == 0
    ret += robot(f | (1 << (x + 4 * (y - 1))), x, y - 1)
  end
  if x < 3 && f & (1 << (x + 1 + 4 * y)) == 0
    ret += robot(f | (1 << (x + 1 + 4 * y)), x + 1, y)
  end
  if y < 3 && f & (1 << (x + 4 * (y + 1))) == 0
    ret += robot(f | (1 << (x + 4 * (y + 1))), x, y + 1)
  end
  ret
end

puts robot(1, 0, 0)
